mat/triangular.go

```diff
@@ -452,15 +452,49 @@ func (t *TriDense) MulTri(a, b Triangular) {
 		defer restore()
 	}
 
-	// TODO(btracey): Improve the set of fast-paths.
+	var aDiag, bDiag bool
+	if aa, ok := aU.(Diagonal); ok {
+		aDiag = true
+		akl, aku := aa.Bandwidth()
+		if akl != 0 || aku != 0 {
+			panic(ErrShape)
+		}
+	}
+	if bb, ok := bU.(Diagonal); ok {
+		bDiag = true
+		bkl, bku := bb.Bandwidth()
+		if bkl != 0 || bku != 0 {
+			panic(ErrShape)
+		}
+	}
+	// If they are both diagonal only need 1 loop.
+	// All diagonal matrices are Upper.
+	if aDiag && bDiag {
+		t.Zero()
+		for i := 0; i < n; i++ {
+			t.SetTri(i, i, a.At(i, i)*b.At(i, i))
+		}
+		return
+	}
+
+	// Now we know at least one matrix is non-diagonal.
+	// And all diagonal matrices are all Upper.
+	// The both-diagonal case is handled above.
```

Comment:
Could you add a TODO here as well for the `Dense` cases as well? These can be much faster if everything is `Dense`, as we can use `Scal` / `Dot`

Reply:
added TODOs and removed the bandwidth checks.

tridense doesn't implement dense etc.  so the straightforward dot(rowview(),colview()) is a larger piece of work.  not sure there is a way to do this without getting too far afield.